src/slice.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SLICE_H
+#define BITCOIN_SLICE_H
+
+#include <type_traits>
+#include <cstddef>
+
+/** A Slice is a pointer to a non-owned continuous array that acts like a container. */
+template<typename C>
+class Slice
+{
+    C* m_data;
+    std::size_t m_size;
+
+public:
+    Slice() : m_data(nullptr), m_size(0) {}
+    Slice(C* data, std::size_t size) : m_data(data), m_size(size) {}
+
+    C* data() const { return m_data; }
+    std::size_t size() const { return m_size; }
+
+    C* begin() const { return m_data; }
+    C* end() const { return m_data + m_size; }
+
+    C& operator[](std::size_t pos) const { return m_data[pos]; }
+};
+
+/** Create a slice to a container exposing data() and size()
+ *
+ * This correctly deals with constness: if either the element type is const
+ * or a reference to a const container is passed, the result will be a const
+ * Slice. */
+template<typename V>
+Slice<typename std::remove_pointer<decltype(std::declval<V>().data())>::type> MakeSlice(V& v) { return Slice<typename std::remove_pointer<decltype(std::declval<V>().data())>::type>(v.data(), v.size()); }
```

Comment:
This could use an overload for the ```std::is_trivially_copyable``` trait that creates a slice for ```Slice<uint8_t>(&v, sizeof(v))```:
> Objects of trivially-copyable types are the only C++ objects that may be safely copied with std::memcpy or serialized to/from binary files with std::ofstream::write()/std::ifstream::read(). In general, a trivially copyable type is any type for which the underlying bytes can be copied to an array of char or unsigned char and into a new object of the same type, and the resulting object would have the same value as the original.

That would eliminate the need for the hack in netaddress.h.